File: android/src/main/kotlin/io/github/sceneview/sceneview_flutter/ARSceneViewConfig.kt
package io.github.sceneview.sceneview_flutter

import com.google.ar.core.Config

class ARSceneViewConfig(
    val lightEstimationMode: Config.LightEstimationMode,
    val instantPlacementMode: Config.InstantPlacementMode,
    val depthMode: Config.DepthMode,
    val planeRenderer: PlaneRenderer,
) {
    companion object {
        fun from(map: Map<String, Any>): ARSceneViewConfig {
            return ARSceneViewConfig(
                Config.LightEstimationMode.values()[map["lightEstimationMode"] as Int],
                Config.InstantPlacementMode.values()[map["instantPlacementMode"] as Int],
                Config.DepthMode.values()[map["depthMode"] as Int],
                PlaneRenderer.from(map["planeRenderer"] as Map<String, Any>)
            )
        }
    }

    override fun toString(): String {
        return "PlaneRenderer: ${planeRenderer}\n" +
                "instantPlacementMode: $instantPlacementMode\n" +
                "lightEstimationMode: $lightEstimationMode\n" +
                "depthMode: $depthMode\n"
    }
}

File: lib/arsceneview_config.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:sceneview_flutter/depth_mode.dart';
import 'package:sceneview_flutter/instant_placement_mode.dart';
import 'package:sceneview_flutter/light_estimation_mode.dart';
import 'package:sceneview_flutter/plane_renderer.dart';

part 'arsceneview_config.freezed.dart';

part 'arsceneview_config.g.dart';

@freezed
class ARSceneviewConfig with _$ARSceneviewConfig {
  const factory ARSceneviewConfig({
    @Default(PlaneRenderer()) PlaneRenderer? planeRenderer,
    @JsonEnum()
    @Default(LightEstimationMode.disabled)
    // @LightEstimationModeConverter()
    LightEstimationMode lightEstimationMode,
    @JsonEnum() @Default(DepthMode.disabled) DepthMode depthMode,
    @JsonEnum() @Default(InstantPlacementMode.disabled) InstantPlacementMode instantPlacementMode,
  }) = _ARSceneviewConfig;

  factory ARSceneviewConfig.fromJson(Map<String, dynamic> json) =>
      _$ARSceneviewConfigFromJson(json);
}
